File: script/cases/dmi.py
#!/usr/bin/env python
#coding=utf-8

case = {
    "id": "DMI",
    "cname": "趋向指标",
    "type": "SUB",
    "src": """
//该模型仅仅用来示范如何根据指标编写简单的模型
//用户需要根据自己交易经验，进行修改后再实际应用!!!
// //后为文字说明，编写模型时不用写出
TR := SUM(MAX(MAX(HIGH-LOW,ABS(HIGH-REF(CLOSE,1))),ABS(LOW-REF(CLOSE,1))),N);//最高价与最低价做差，最高价与前一周期收盘价做差，最低价与前一周期收盘价作差，在上述三个数据中取绝对值最大者，对该最大值做N周期累加求和。。
HD := HIGH-REF(HIGH,1);//最高价与前一周期最高价做差
LD := REF(LOW,1)-LOW;//前一周期最低价与最低价做差
DMP:= SUM(IFELSE(HD>0 && HD>LD,HD,0),N);//如果HD>0并且HD>LD,取HD否则取0,对取值做N周期累加求和。
DMM:= SUM(IFELSE(LD>0 && LD>HD,LD,0),N);//如果LD>0并且LD>HD,取LD否则取0,对取值做N周期累加求和。
PDI: DMP*100/TR;
MDI: DMM*100/TR;
ADX: MA(ABS(MDI-PDI)/(MDI+PDI)*100,M);//MDI与PDI差的绝对值与(MDI+PDI)*100做比值，取该比值的M个周期均值。
ADXR:(ADX+REF(ADX,M))/2;
BACKGROUNDSTYLE(1);
""",
    "params": [
        ("N", 1, 100, 14),
        ("M", 1, 100, 6),
        ("N0", 1, 100, 20),
        ("M0", 1, 100, 20),
    ],
    "expected": """
    """,
}

File: src/libs/ind/dmi.js
function* dmi(C){
    //指标定义
    C.DEFINE({
        type: "SUB",
        cname: "趋向指标",
        state: "KLINE"
    });
    let n = C.PARAM(14, "N", {min:1, max:100});
    let m = C.PARAM(6, "M", {min:1, max:30});
    //输出序列
    let pdi = C.OUTS("LINE", "PDI", {color: YELLOW});
    let mdi = C.OUTS("LINE", "MDI", {color: GREEN});
    let adx = C.OUTS("LINE", "ADX", {color: LIGHTBLUE});
    let adxr = C.OUTS("LINE", "ADXR", {color: LIGHTRED});

    let tr = [];
    let sum_tr = [];
    let dmp = [];
    let sum_dmp = [];
    let dmm =[];
    let sum_dmm = [];
    let ad = [];
    //计算
    while(true) {
        let i = yield;
        let h = C.DS.high[i];
        let l = C.DS.low[i];
        let last_c = C.DS.close[i-1];
        tr[i] = MAX( h-l, Math.abs(last_c-h), Math.abs(last_c-l));
        sum_tr[i] = SUM(i, tr, n, sum_tr);
        let hd = h - C.DS.high[i-1];
        let ld = C.DS.low[i-1] - l;
        dmp[i] = hd>0 && hd>ld ? hd: 0;
        sum_dmp[i] = SUM(i, dmp, n, sum_dmp);
        dmm[i] = ld>0 && ld >hd ? ld : 0;
        sum_dmm[i] = SUM(i, dmm, n, sum_dmm);

        pdi[i] = sum_dmp[i] / sum_tr[i] * 100;
        mdi[i] = sum_dmm[i] / sum_tr[i] * 100;
        ad[i] = Math.abs(mdi[i]-pdi[i]) / (mdi[i]+pdi[i]) * 100;
        adx[i] = MA(i, ad, m, adx);
        adxr[i] = (adx[i] + adx[i-m+1]) / 2;
    }
}
